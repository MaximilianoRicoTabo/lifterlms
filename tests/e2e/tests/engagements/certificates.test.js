import {
	insertBlock,
	switchUserToAdmin,
	visitAdminPage
} from '@wordpress/e2e-test-utils';

import {
	clearBlocks,
	clickAndWait,
	clickElementByText,
	createCertificate,
	fillField,
	getAllBlocks,
	loginStudent,
	logoutUser,
	openSidebarPanelTab,
	registerStudent,
	toggleOpenRegistration,
	toggleSidebarPanel,
	visitPostPermalink,
} from '@lifterlms/llms-e2e-test-utils';


/**
 * Retrieves the HTML for the currently-viewed certificate.
 *
 * Modifies the DOM to replace IDs from html wrapper attributes with a mocked ID, "9999".
 *
 * Also replaces the "Awarded" date with a mocked date, "Octember 01, 9999".
 *
 * @since [version]
 *
 * @param {Number} templateVersion The editor template version of the certificate.
 * @return {string} The outerHTML of the certificate element.
 */
async function getCertificateHTML( templateVersion = 2 ) {

	let WRAPPER = '.llms-certificate-wrapper';

	if ( 1 === templateVersion ) {
		WRAPPER = '.llms-certificate-container';
	}

	await page.waitForSelector( WRAPPER );

	let html = await page.$eval( WRAPPER, el => el.outerHTML );

	// Hardcode IDs for the snapshot.
	html = html.replace( /(id="certificate-)\d+(")/, '$1999$2' ); // ID attribute.
	html = html.replace( /( post-)\d+( )/, '$1999$2' ); // ID classname.

	// Replace the awarded date with a mocked date.
	html = html.replace( /(">On ).+(<\/p>)/, '$1Octember 01, 9999$2' );

	// Strip the classname added by the various themes; 2022 theme doesn't include it and this makes it so we can have a single snapshot.
	html = html.replace( ' entry">', '">' );

	return html;

}

/**
 * Retrieves an elements CSS inline css rules a JS object.
 *
 * @since [version]
 *
 * @param {string} selector The DOM query selector string.
 * @return {Object} An object of css rules.
 */
async function getElementStyles( selector ) {

	return await page.$eval( selector, ( { style } ) => {

		const styles = {};

	    for ( let i = 0; i < style.length; i++ ) {
	        let item = style.item( i );
	        styles[ item ] = style[ item ];
	    }

	    return styles;

	} );

}

describe( 'Engagements/Certificates', () => {

	beforeAll( async () => {
		await switchUserToAdmin();
		await toggleOpenRegistration( true );
	} );

	afterAll( async () => {
		await switchUserToAdmin();
		await toggleOpenRegistration( false );
	} );

	describe( 'Templates', () => {

		it ( 'should create a certificate in the block editor', async () => {

			await createCertificate( {
				title: 'A Certificate!',
			} );

			await visitPostPermalink();
			expect( await getCertificateHTML() ).toMatchSnapshot();

		} );

		it ( 'can reset blocks to the default template', async () => {

			await visitAdminPage( 'post-new.php', 'post_type=llms_certificate' );

			await page.waitForTimeout( 1000 );

			// Add a title.
			const TITLE_SELECTOR = '.is-root-container.block-editor-block-list__layout .wp-block-llms-certificate-title';
			await page.waitForSelector( TITLE_SELECTOR );
			await page.click( TITLE_SELECTOR );
			await page.keyboard.type( 'Certificate Title' );
			await page.keyboard.press( 'Tab' );

			// Add a new block.
			await insertBlock( 'Spacer' );

			// Make sure sidebar is open.
			await openSidebarPanelTab();

			// Open confirmation modal.
			await clickElementByText( 'Reset template', '.edit-post-post-status button' );

			// Confirm reset.
			await clickElementByText( 'Reset template', '.components-modal__frame button' );

			await page.waitForTimeout( 1000 );

			expect( await getAllBlocks( false ) ).toMatchSnapshot();

		} );

		it ( 'can reset blocks to the default template when there are no blocks present', async () => {

			await visitAdminPage( 'post-new.php', 'post_type=llms_certificate' );

			await clearBlocks();

			// Make sure sidebar is open.
			await openSidebarPanelTab();

			// Open confirmation modal.
			await clickElementByText( 'Reset template', '.edit-post-post-status button' );

			// Confirm reset.
			await clickElementByText( 'Reset template', '.components-modal__frame button' );

			await page.waitForTimeout( 1000 );

			expect( await getAllBlocks( false ) ).toMatchSnapshot();

		} );

		it ( 'should update the DOM when certificate settings change', async () => {

			await visitAdminPage( 'post-new.php', 'post_type=llms_certificate' );
			await toggleSidebarPanel( 'Settings' );

			const assertStylesMatch = async ( hint = null ) => {

				expect(
					await getElementStyles( '.is-root-container.block-editor-block-list__layout' )
				).toMatchSnapshot(
					{ 'background-image': expect.any( String ) },
					hint
				);

			};

			const testSize = async ( orientation, sizeName ) => {

				await clickElementByText( orientation, '.llms-certificate-doc-settings button' );
				await page.waitForTimeout( 500 );

				await assertStylesMatch( `Size: ${ sizeName } (${ orientation })` );

			};

			// Test registered sizes.
			const { sizes } = await page.evaluate( () => window.llms.certificates ),
				sizeKeys = Object.keys( sizes );

			const SIZE_SELECTOR = '#llms-certificate-control--size';
			await page.waitForSelector( SIZE_SELECTOR );

			// Loop through all available sizes.
			for ( let i = 0; i < sizeKeys.length; i++ ) {

				const { name } = sizes[ sizeKeys[ i ] ];

				await page.select( SIZE_SELECTOR, sizeKeys[ i ] );

				await testSize( 'Landscape', name );
				await testSize( 'Portrait', name );

			}

			// Custom size
			await page.select( SIZE_SELECTOR, 'CUSTOM' );
			await fillField( '#llms-certificate-control--size--custom-width', '5.5' );
			await fillField( '#llms-certificate-control--size--custom-height', '10' );
			await testSize( 'Landscape', 'CUSTOM' );
			await testSize( 'Portrait', 'CUSTOM' );


			// Margins.
			await fillField( '#llms-certificate-control--margin--top', '2.5' );
			await fillField( '#llms-certificate-control--margin--right', '3' );
			await fillField( '#llms-certificate-control--margin--bottom', '7' );
			await fillField( '#llms-certificate-control--margin--left', '10' );
			await assertStylesMatch( 'Margins' );

			// Background color chooser.
			const colors = await page.$$( '.components-circular-option-picker .components-circular-option-picker__option' );

			for ( let j = 0; j < colors.length; j++ ) {

				// The color of the button we're pressing.
				const expectedBgColor = await colors[ j ].evaluate( ( { style } ) => style['background-color'] );

				// Press it.
				await colors[ j ].click();
				await page.waitForTimeout( 500 );

				// Get the updated background color of the editor wrapper.
				const styles = await getElementStyles( '.editor-styles-wrapper' ),
					bgColor = styles['background-color'];

				expect( bgColor ).toBe( expectedBgColor );

			}

		} );

		// it ( 'should be able to award a certificate directly to a student', async () => {} );



		// it ( 'should create a certificate in the classic editor', async () => {

		// 	// Required due to WP core bug in 5.6.1 & later, see https://core.trac.wordpress.org/ticket/52440.
		// 	page.on( 'dialog', dialog => dialog.accept() );

		// 	await visitAdminPage( 'post.php', `post=${ certificateId }&action=edit` );
		// 	await visitPostPermalink();
		// 	expect( await getCertificateHTML() ).toMatchSnapshot();

		// } );


		// it ( 'should be able to view and delete an earned certificate from reporting screens', async () => {

		// 	// Create a user who will earn the certificate.
		// 	const
		// 		first     = 'Student',
		// 		last      = 'WithACert',
		// 		{ email } = await registerStudent( { first, last } );
		// 	await logoutUser();

		// 	await visitAdminPage( 'users.php', `s=${ encodeURIComponent( email ) }` );

		// 	await page.goto( await page.$eval( '#the-list tr:first-child span.llms-reporting a', el => `${ el.href }&stab=certificates` ) );

		// 	const reportingUrl = await page.url();

		// 	// Navigate to the certificate page.
		// 	await page.goto( await page.$eval( '#llms-gb-table-certificates td.actions a', el => el.href ) );

// expect( await getCertificateHTML() ).toMatchSnapshot();

		// 	await page.goto( reportingUrl );
		// 	// page.on( 'dialog', dialog => dialog.accept() ); // Uncomment when https://core.trac.wordpress.org/ticket/52440 is resolved.
		// 	await clickAndWait( '#llms_delete_cert' );

		// } );

	} );

	describe( 'Awarded', () => {

		it ( 'should award a certificate on user registration', async () => {

			await createCertificate( {
				title: 'Awarded on Registration',
				engagement: 'user_registration',
			} );

			const
				first = 'Maude',
				last  = 'Lebowski',
				user  = await registerStudent( { first, last } );

			await page.waitForTimeout( 1000 );

			// Certificate listed on the certs area of the dashboard.
			const selector = '.llms-sd-section.llms-my-certificates li.llms-certificate-loop-item a.llms-certificate';
			expect( await page.$eval( `${ selector } h4.llms-certificate-title`, el => el.textContent ) ).toBe( 'Awarded on Registration' );

			// Visit the certificate permalink.
			await clickAndWait( selector );
			const certUrl = await page.url();

			// View the cert.
			expect( await getCertificateHTML() ).toMatchSnapshot();

			// Logged out user cannot view.
			await logoutUser();
			await page.goto( certUrl );
			expect( await page.waitForSelector( 'body.error404' ) ).toBeTruthy();

			// Student can enable sharing.
			await loginStudent( user.email, user.pass );
			await page.goto( certUrl );

			// Enable sharing.
			await clickAndWait( 'button[name="llms_enable_cert_sharing"]' );

			// Logged out user can view.
			await logoutUser();
			await page.goto( certUrl );
			expect( await getCertificateHTML() ).toMatchSnapshot();
			await expect( await page.$( '#llms-print-certificate' ) ).toBeNull();

		} );

	} );

} );
