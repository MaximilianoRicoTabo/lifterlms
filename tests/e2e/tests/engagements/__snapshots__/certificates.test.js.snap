// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engagements/Certificates Awarded should award a certificate on user registration 1`] = `
"<div class=\\"llms-certificate-wrapper\\">
	<div id=\\"certificate-999\\" class=\\"llms-certificate-container cert-template-v2 post-999 llms_my_certificate type-llms_my_certificate status-publish hentry entry\\">

		
		
		
<h1 class=\\"has-text-align-center has-default-font-family\\" id=\\"awarded-on-registration\\">Awarded on Registration</h1>



<p class=\\"has-text-align-center\\">Awarded to Maude Lebowski</p>



<p class=\\"has-text-align-center\\">On Octember 01, 9999</p>

		
	</div>
</div>"
`;

exports[`Engagements/Certificates Awarded should award a certificate on user registration 2`] = `
"<div class=\\"llms-certificate-wrapper\\">
	<div id=\\"certificate-999\\" class=\\"llms-certificate-container cert-template-v2 post-999 llms_my_certificate type-llms_my_certificate status-publish hentry entry\\">

		
		
		
<h1 class=\\"has-text-align-center has-default-font-family\\" id=\\"awarded-on-registration\\">Awarded on Registration</h1>



<p class=\\"has-text-align-center\\">Awarded to Maude Lebowski</p>



<p class=\\"has-text-align-center\\">On Octember 01, 9999</p>

		
	</div>
</div>"
`;

exports[`Engagements/Certificates Templates should create a certificate in the block editor 1`] = `
"<div class=\\"llms-certificate-wrapper\\">
	<div id=\\"certificate-999\\" class=\\"llms-certificate-container cert-template-v2 post-999 llms_certificate type-llms_certificate status-publish hentry entry\\">

		
		
		
<h1 class=\\"has-text-align-center has-default-font-family\\" id=\\"a-certificate\\">A Certificate!</h1>



<p class=\\"has-text-align-center\\">Awarded to Chad Feldheimer</p>



<p class=\\"has-text-align-center\\">On Octember 01, 9999</p>

		
	</div>
</div>"
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Background Color 1`] = `"rgb(0, 0, 0)"`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Background Color 2`] = `"rgb(40, 48, 61)"`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Background Color 3`] = `"rgb(57, 65, 77)"`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Background Color 4`] = `"rgb(209, 228, 221)"`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Background Color 5`] = `"rgb(209, 223, 228)"`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Background Color 6`] = `"rgb(209, 209, 228)"`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Background Color 7`] = `"rgb(228, 209, 209)"`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Background Color 8`] = `"rgb(228, 218, 209)"`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Background Color 9`] = `"rgb(238, 234, 221)"`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Background Color 10`] = `"rgb(255, 255, 255)"`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Margins 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "5.5in 10in",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "10in",
  "padding-bottom": "7%",
  "padding-left": "10%",
  "padding-right": "3%",
  "padding-top": "2.5%",
  "width": "5.5in",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: A3 (Landscape) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "420mm 297mm",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "297mm",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "420mm",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: A3 (Portrait) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "297mm 420mm",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "420mm",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "297mm",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: A4 (Landscape) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "297mm 210mm",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "210mm",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "297mm",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: A4 (Portrait) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "210mm 297mm",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "297mm",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "210mm",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: A5 (Landscape) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "210mm 148mm",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "148mm",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "210mm",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: A5 (Portrait) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "148mm 210mm",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "210mm",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "148mm",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: CUSTOM (Landscape) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "10in 5.5in",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "5.5in",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "10in",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: CUSTOM (Portrait) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "5.5in 10in",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "10in",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "5.5in",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: Ledger (Landscape) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "17in 11in",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "11in",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "17in",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: Ledger (Portrait) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "11in 17in",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "17in",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "11in",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: Legal (Landscape) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "14in 8.5in",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "8.5in",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "14in",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: Legal (Portrait) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "8.5in 14in",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "14in",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "8.5in",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: Letter (Landscape) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "11in 8.5in",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "8.5in",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "11in",
}
`;

exports[`Engagements/Certificates Templates should update the DOM when certificate settings change: Size: Letter (Portrait) 1`] = `
Object {
  "background-image": Any<String>,
  "background-repeat-x": "no-repeat",
  "background-repeat-y": "no-repeat",
  "background-size": "8.5in 11in",
  "box-sizing": "border-box",
  "margin-left": "auto",
  "margin-right": "auto",
  "min-height": "11in",
  "padding-bottom": "5%",
  "padding-left": "5%",
  "padding-right": "5%",
  "padding-top": "5%",
  "width": "8.5in",
}
`;
